### ISSUE TITLE
A way to setup custom Control's theme options

### ISSUE THREAD

**Issue by ghsoares** [godotengine/godot-proposals#3257]
### Describe the project you are working on

I'm creating a tool for reactive game user interface design.

### Describe the problem or limitation you are having in your project

In my project I have some custom Control nodes that can be used on multiple projects, like a ButtonContainer, that fits to the children's rect and still can be used as a button. The problem is that I want to tell the editor which styles this custom node uses and can be overridden, eg. ButtonContainer node has the same styles of a Button, pressed, normal, hover, disabled and focus, but there is no way to tell Godot that the node uses theses named styles.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

The feature is to add a custom virtual function called `_get_stylebox_list()`, similar to Theme's `get_stylebox_list(node_type)` but you can set the theme styles list directly into the custom node. Also add another virtual function called `_get_stylebox_default(style_name)` which returns the default StyleBox used by this custom node with the name `style_name`.
This could be applied for constants, colors, icons, etc. Provide a function that returns a array and a function that returns the default.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

for `_get_stylebox_list`, you can override and return a array of strings, like this:
```gdscript
func _get_stylebox_list() -> Array:
    return ["normal", "pressed", "hover", "disabled", "focus"]
```

And then for `_get_stylebox_default` you can override and return a StyleBox based on `style_name`, like this:

```gdscript
func _get_stylebox_default(style_name) -> StyleBox:
    var style := StyleBoxFlat.new()
    if style_name == "normal":
        style.bg_color = Color.red 
    elif style_name == "pressed":
        style.bg_color = Color.blue
    #and so on
    return style
```

Then if you go to the node's inspector, the `theme_overrides` would show these styles to override.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No, because it needs to change Control's core code

### Is there a reason why this should be core and not an add-on in the asset library?

Same as above

**Comment by KoBeWi** [comment]
You can easily define arbitrary types and theme items in a Theme using the new theme editor:
![image](https://user-images.githubusercontent.com/2223172/132135096-e76392b5-a08c-4cf6-902c-e2ab2ebbbae7.png)
Any type/item you define this way will appear under Theme resource in the inspector too and you can access them from script using e.g. `get_stylebox("my_stylebox1", "MyCustomType")`.

**Comment by ghsoares** [comment]
@KoBeWi What I'm proposing is not for arbitrary types, but for custom types defined with `class_name`. The way you are showing there's no definitive way to tell which styles this custom node have and if the style name is correct, I would like to view then and override then in the inspector, not in a Theme resource.

**Comment by KoBeWi** [comment]
Well if you want custom properties why don't you just use `export my_stylebox: StyleBox`? They won't show the same way as for theme styles, but it shouldn't be a problem.

**Comment by pycbouh** [comment]
> @KoBeWi What I'm proposing is not for arbitrary types, but for custom types defined with `class_name`. The way you are showing there's no definitive way to tell which styles this custom node have and if the style name is correct, I would like to view then and override then in the inspector, not in a Theme resource.

Unfortunately, that's how all controls work, even the built-ins. There is no contract as to what theme properties controls would require. The list of local overrides is generated from the properties defined on the default theme for that built-in class. You can check `Control::_get_property_list` to see what is going on.

Now, I don't think your proposed implementation is good. We don't just have styleboxes, we have 6 different theme item types, so that's a lot of individual methods. But having strictly defined contracts is what I wanted for controls for a long time. It would be something like you describe, but closer to `_get_property_list`. As for default values, reduz wanted to solve this by introducing theme extensions that would be automatically added to the default theme by scanning the file system. With that it should be possible to provide custom controls that are as flexible as built-ins.

But that's not in works yet, so you would have to do `MyCustomType` thing one way or another, or export the necessary properties manually (with `export`, like KoBeWi says, or with `_get_property_list`, like the `Control` does).

**Comment by Drako** [comment]
> Unfortunately, that's how all controls work, even the built-ins. There is no contract as to what theme properties controls would require. The list of local overrides is generated from the properties defined on the default theme for that built-in class. You can check Control::_get_property_list to see what is going on.

That is not true, even though I'm not sure if it was at the time of your comment.

Right now the process for the builtin controls is as follows:
 * Like all builtin classes accessible via GDScript/C#, controls have a `_bind_methods` function, which specifies, which functions, properties etc. are exposed to scripts.
 * For methods, properties etc. they call `ClassDB` methods like `bind_method`, `add_property`, ...
 * Controls additionally call the `ThemeDB` method `bind_class_item`, which takes care of registering the theme property.
 * Most controls don't directly call `ThemeDB::bind_class_item`, but instead use one of the macros `BIND_THEME_ITEM` or `BIND_THEME_ITEM_CUSTOM` (the latter one, if the theme property and the C++ attribute don't share the same name).

As the most trivial example, `Panel`, which only has a stylebox, does the following:
```cpp
void Panel::_bind_methods() {
	BIND_THEME_ITEM_CUSTOM(Theme::DATA_TYPE_STYLEBOX, Panel, panel_style, "panel");
}
```

The issue is, that while `ThemeDB` is accessible by scripts, `bind_class_item` is not exposed.
Now making `bind_class_item` available to scripts seems trivial, but I have a feeling that this would not be sufficient to make the ThemeEditor pick it up.

I also found, that `ThemeDB` was apparently removed from the GDExtension API.

So right now the only way to have custom controls with real theme properties is writing your own module which requires compiling the engine from source.

**Comment by Drako** [comment]
It should also be noted, that `_bind_methods` does not exist for scripts.

In GDScript, every method and every property is available to other scripts implicitly.
Special stuff like properties being exposed in the Inspector are handled via annotations.

In C# everything is accessible according to their visibility to other C# scripts.
Special stuff is also handled via annotations.
GDScript <-> C# Interop does have some pitfalls.

So I think the most appropriate design would not be exposing `ThemeDB`, but instead adding an annotation.
My idea would be something like `@export_theme` or `[Export(PropertyHint.Theme)]` respectively, to mark a property as a theme property and also show it in the Inspector under the "Theme Overrides".

But I'm open for other name suggestions, if people feel like this doesn't belong to "export".

I'd be happy to implement that feature though :)

**Comment by YuriSizov** [comment]
> That is not true, even though I'm not sure if it was at the time of your comment.

Yes, I implemented everything that you just described since I had made that comment. This work is tracked in https://github.com/godotengine/godot-proposals/issues/4486. Exposing the system to scripting was the next step, but unfortunately I will never be able to complete this work. So if you want to give it go, feel free.

One tip though. You cannot use export annotations for this because theme item are not class properties. They are special, and the logic that is required needs to be special too. For instance, it's unclear what should happen if you try to access or copy a member like that in an external script. I think they should be more akin to signals in this regard.

You also need to consider how default values are treated. Built-in nodes don't have a way to set defaults, instead defaults are provided by the default theme. But scripted nodes must be able to provide defaults, so the theme items that you declare there must provide default values. At the same time, these theme items in scripts should not be confused with local overrides.

So you can see how this can get complicated, and a bespoke solution is required. It needs to be coordinated with scripting teams and the extension team.

**Comment by Drako** [comment]
> but unfortunately I will never be able to complete this work

That sounds a bit worrying, I hope you are doing fine.
I'll have a look at your linked issue later, when I get the time.

Off-topic story:
I like to do things in the intended clean way. I'm currently following a Godot 3D course and in one section it is explained how to do custom Controls with their own `_draw` method. Now that introduction is very rudimentary and from previous experiments with Themes I thought to myself... "well, wouldn't it make sense, to get the drawing color from the theme?" and similar things. This stuff kinda works if you make a custom button and just inherit the theming options from Button, but if you make a completely custom Control, you also need custom properties.
So I got into a research rabbit hole of looking through the documentation on Control, Theme and looking for ThemeDB, which is inaccessible in GDScript, but accessible in C#, but only with a limited set of functionality.
I came across this proposal and then went into the engine source code to see how stuff is done there,
just to realize, that currently there is no way to actually make a custom control that behaves and integrates as smoothly as a builtin one. Not from an addon, not from a GDExtension, not from a regular script.

> I implemented everything that you just described since I had made that comment.

also this is awesome :) maybe I might be able to finish your work.

**Comment by KoBeWi** [comment]
You can custom types to your Theme. The only difference from regular Controls is that you won't see overrides in the inspector.

**Comment by Drako** [comment]
That's not the same and you know it.

If you make a plugin with custom controls, you can hardly expect the user to go into the generic theme editor to manually add your Controls and their theme properties. That's not how good UX works.

Also these "only differences" are literally what this GIP is about.

I don't want to come off as rude, but your comment was more appropriate for a beginner question of the kind "hey, how can I do X?" and you'd be like "you can't, but you can do Y, which is pretty similar".
However it is not a useful comment for a discussion, where it is already established, that X is not possible, where it is already established, that Y is possible, and where the entire point is making X possible, because of the limitations of Y.

**Comment by Drako** [comment]
The thing is, we want the ability to make custom controls that behave like regular ones.
For one it feels nicer if you have consistency. Also AddOn authors would be able to make custom controls, which appear in the Theme editor and which have all their adjustable theme properties displayed.

The things you suggest don't solve that. Yes, if I make my own little game, I can deal with being dissatisfied, because I can't have clean custom controls. But for AddOn authors this is a question of UX. And a great AddOn has great UX.

Whenever you recommended something, it always came with an "only difference", but it's literally these differences that we find problematic.
